ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Ledger.hs

```diff
@@ -519,71 +522,242 @@ data HardForkLedgerWarning xs =
+     CanHardFork xs
+  => NP WrapPartialLedgerConfig xs
+  -> NP (K EraParams) xs
+  -> NP TopLevelConfig xs
+  -> NS (Current LedgerState) xs
+  -> NS (Current LedgerState) xs
+  -> [LedgerEvent (HardForkBlock xs)]
+inspectHardForkLedger = go
   where
-    aux :: LedgerWarning blk -> HardForkLedgerWarning xs
-    aux = HardForkLedgerWarningInEra
-        . OneEraLedgerWarning
-        . unK
-        . apFn inj
-        . WrapLedgerWarning
-
-additionalChecks :: SingleEraBlock blk
-                 => WrapPartialLedgerConfig blk
-                 -> K EraParams blk
-                 -> Current LedgerState blk
-                 -> K [HardForkLedgerWarning xs] blk
-additionalChecks (WrapPartialLedgerConfig pcfg) (K eraParams) Current{..} = K $
-    concat [
-        [ HardForkLedgerWarningUnexpectedTransition eraParams transition
-        | Just transition <- [singleEraTransition
-                                pcfg
-                                eraParams
-                                currentStart
-                                currentState]
-        , not $ validLowerBound
-                  (History.safeBeforeEpoch (History.eraSafeZone eraParams))
+    go :: All SingleEraBlock xs
+       => NP WrapPartialLedgerConfig xs
+       -> NP (K EraParams) xs
+       -> NP TopLevelConfig xs
+       -> NS (Current LedgerState) xs
+       -> NS (Current LedgerState) xs
+       -> [LedgerEvent (HardForkBlock xs)]
+
+    go (pc :* _) (K ps :* pss) (c :* _) (Z before) (Z after) = concat [
+          map liftEvent $
+            inspectLedger c (currentState before) (currentState after)
+
+        , case (pss, confirmedBefore, confirmedAfter) of
+            (_, Nothing, Nothing) ->
+              []
+            (_, Just _, Nothing) ->
+              return $ LedgerWarning $
+                HardForkWarningTransitionUnconfirmed eraIndexZero
+            (Nil, Nothing, Just transition) ->
+              return $ LedgerWarning $
+                HardForkWarningTransitionInFinalEra eraIndexZero transition
+            (Nil, Just transition, Just transition') -> do
+              -- Only warn if the transition has changed
+              guard (transition /= transition')
+              return $ LedgerWarning $
+                HardForkWarningTransitionInFinalEra eraIndexZero transition
+            ((:*){}, Nothing, Just transition) ->
+              return $
+                if validLowerBound (History.safeBeforeEpoch safeZone) transition
+                  then LedgerUpdate $
+                         HardForkUpdateTransitionConfirmed
+                           eraIndexZero
+                           (eraIndexSucc eraIndexZero)
+                           transition
+                  else LedgerWarning $
+                         HardForkWarningTransitionMismatch
+                           eraIndexZero
+                           ps
+                           transition
+            ((:*){}, Just transition', Just transition) -> do
```

Comment:
Is `transition'`, `transition` (note the `'`) intentional?

Reply:
No, well-spotted.